src/order/circular.lean

```diff
@@ -0,0 +1,387 @@
+The fields are propeq, but not defeq.
+
+## TODO
+
+Antisymmetry is quite weak in the sense that there's no way to discriminate which two points are
+equal. This prevents defining closed-open intervals `cIco` and `cIoc` in the neat `=`-less way. We
+currently haven't defined them at all.
+
+What's next is to define circular groups and provide instances for `zmod` and the usual circle group
+`ℝ/ℤ`.
+
+## References
+
+* https://en.wikipedia.org/wiki/Cyclic_order
+* https://en.wikipedia.org/wiki/Partial_cyclic_order
+
+## Tags
+
+circular order, cyclic order, circularly ordered set, cyclically ordered set
+-/
+
+/-- Syntax typeclass for a betweenness relation. -/
+class has_btw (α : Type*) :=
+(btw : α → α → α → Prop)
+
+export has_btw (btw)
+
+notation `[` a ` ≤ ` b ` ≤ ` c `]` := btw a b c
+
+/-- Syntax typeclass for a strict betweenness relation. -/
+class has_sbtw (α : Type*) :=
+(sbtw : α → α → α → Prop)
+
+export has_sbtw (sbtw)
+
+notation `[` a ` < ` b ` < ` c `]` := sbtw a b c
+
+/-- A circular preorder is the analogue of a preorder where you can loop around. `≤` and `<` are
+replaced by ternary relations `btw` and `sbtw`. `btw` is reflexive and cyclic. `sbtw` is transitive.
+-/
+class circular_preorder (α : Type*) extends has_btw α, has_sbtw α :=
+(btw_refl (a : α) : btw a a a)
+(btw_cyclic_left {a b c : α} : btw a b c → btw b c a)
+(sbtw := λ a b c, btw a b c ∧ ¬btw c b a)
+(sbtw_iff_btw_not_btw {a b c : α} : sbtw a b c ↔ (btw a b c ∧ ¬btw c b a) . order_laws_tac)
+(sbtw_trans_left {a b c d : α} : sbtw a b c → sbtw b d c → sbtw a d c)
+
+export circular_preorder (btw_refl) (btw_cyclic_left) (sbtw_trans_left)
+
+/-- A circular partial order is the analogue of a partial order where you can loop around. `≤` and
+`<` are replaced by ternary relations `btw` and `sbtw`. `btw` is reflexive, cyclic and
+antisymmetric. `sbtw` is transitive. -/
+class circular_partial_order (α : Type*) extends circular_preorder α :=
+(btw_antisymm {a b c : α} : btw a b c → btw c b a → a = b ∨ b = c ∨ c = a)
+
+export circular_partial_order (btw_antisymm)
+
+/-- A circular order is the analogue of a linear order where you can loop around. `≤` and `<` are
+replaced by ternary relations `btw` and `sbtw`. `btw` is reflexive, cyclic, antisymmetric and total.
+`sbtw` is transitive. -/
+class circular_order (α : Type*) extends circular_partial_order α :=
+(btw_total : ∀ a b c : α, btw a b c ∨ btw c b a)
+
+export circular_order (btw_total)
+
+/-! ### Circular preorders -/
+
+section circular_preorder
+variables {α : Type*} [circular_preorder α]
+
+lemma btw_rfl {a : α} : btw a a a :=
+btw_refl _
+
+-- TODO: `alias` creates a def instead of a lemma.
+-- alias btw_cyclic_left        ← has_btw.btw.cyclic_left
+lemma has_btw.btw.cyclic_left {a b c : α} (h : btw a b c) : btw b c a :=
+btw_cyclic_left h
+
+lemma btw_cyclic_right {a b c : α} (h : btw a b c) : btw c a b :=
+h.cyclic_left.cyclic_left
+
+alias btw_cyclic_right ← has_btw.btw.cyclic_right
```

Comment:
I don't get this TODO: you seem to use `alias` anyway.

Reply:
No, `alias btw_cyclic_left        ← has_btw.btw.cyclic_left` fails, but not the other ones. This is because `btw_cyclic_left` is a field of `circular_preorder` and `alias` seems to mishandle those? I don't know.